STAT540 Seminar 05 - Low volume linear modelling
================================================

Rod Docking - 2014-02-05

Running through the material at [seminar_05](http://www.ugrad.stat.ubc.ca/~stat540/seminars/seminar05_lowVolumeLinearModelling.html)

Setup
-----

Load in libraries and data:

```{r}
library(lattice)
library(ggplot2)
library(plyr)
prDat <- read.table("../../stat540_2014/examples/photoRec/data/GSE4051_data.tsv")
str(prDat, max.level = 0)
prDes <- readRDS("../../stat540_2014/examples/photoRec/data/GSE4051_design.rds")
str(prDes)
```

Write a function to prepare a mini-dataset for a small number of genes
----------------------------------------------------------------------

I'll need to write a function to regenerate the given output...

```{r}
(luckyGenes <- c("1419655_at","1438815_at"))
```

Which has expected `str`, `head`, and `tail` of:

```
jDat <- prepareData(luckyGenes)
str(jDat)
## 'data.frame':    78 obs. of  6 variables:
##  $ sidChar : chr  "Sample_20" "Sample_21" "Sample_22" "Sample_23" ...
##  $ sidNum  : num  20 21 22 23 16 17 6 24 25 26 ...
##  $ devStage: Factor w/ 5 levels "E16","P2","P6",..: 1 1 1 1 1 1 1 2 2 2 ...
##  $ gType   : Factor w/ 2 levels "wt","NrlKO": 1 1 1 1 2 2 2 1 1 1 ...
##  $ gExp    : num  10.93 10.74 10.67 10.68 9.61 ...
##  $ gene    : Factor w/ 2 levels "1419655_at","1438815_at": 1 1 1 1 1 1 1 1 1 1 ...
head(jDat)
##     sidChar sidNum devStage gType   gExp       gene
## 1 Sample_20     20      E16    wt 10.930 1419655_at
## 2 Sample_21     21      E16    wt 10.740 1419655_at
## 3 Sample_22     22      E16    wt 10.670 1419655_at
## 4 Sample_23     23      E16    wt 10.680 1419655_at
## 5 Sample_16     16      E16 NrlKO  9.606 1419655_at
## 6 Sample_17     17      E16 NrlKO 10.840 1419655_at
tail(jDat)
##      sidChar sidNum devStage gType  gExp       gene
## 73 Sample_38     38  4_weeks    wt 8.211 1438815_at
## 74 Sample_39     39  4_weeks    wt 8.436 1438815_at
## 75 Sample_11     11  4_weeks NrlKO 8.465 1438815_at
## 76 Sample_12     12  4_weeks NrlKO 8.841 1438815_at
## 77  Sample_2      2  4_weeks NrlKO 8.506 1438815_at
## 78  Sample_9      9  4_weeks NrlKO 8.952 1438815_at
```

What I'm starting from is `prDes` (the descriptions, basically):

```{r}
str(prDes)
head(prDes)
```

In this data frame, *rows* are one for each sample, with *columns* indicating Sample ID number, sample name, developmental stage, and genotype.

And `prDat` (the actual data):

```{r}
str(prDat)
head(prDat, n=2)
```

In this data frame, *rows*, are one for each gene, with columns giving the gene expression value for each of the `r nrow(prDes)` samples.

So, this is going to be similar to what we did in seminar 4. Add the gene expression values for the selected genes, as well as the gene name, onto the `prDat` data frame. We'll need to use recycling to get all the rows properly.

Here's a modification of the seminar 04 code as a function:

```{r}
prepareData <- function(keepGenes){
  miniDat <- subset(prDat, rownames(prDat) %in% keepGenes)
  miniDat <- data.frame(gExp = as.vector(t(as.matrix(miniDat))),
                        gene = factor(rep(rownames(miniDat), 
                                      each = ncol(miniDat)),
                                       levels = keepGenes))
  miniDat <- suppressWarnings(data.frame(prDes, miniDat))
}
```

Try it out:

```{r}
jDat <- prepareData(luckyGenes)
str(jDat)
head(jDat)
tail(jDat)
```

Looks good! For real TDD, I'd match these to the expected results in code, but this is OK for now.

Check by plotting:

```{r}
stripplot(gExp ~ devStage | gene, jDat,
          group = gType, jitter.data = TRUE,
          auto.key = TRUE, type = c('p', 'a'), grid = TRUE)
```

The plot looks similar, although the colours and facetting are different. Most likely some sort of `lattice` option I'm missing/
