Seminar 04 - Two group comparisons and data aggregation
========================================================

Preamble
--------

Load in the datasets:

```{r}
library(lattice)
prDat <- read.table("../../../stat540/stat540_2014/examples/photoRec/data/GSE4051_data.tsv")
str(prDat, max.level = 0)
prDes <- readRDS("../../../stat540/stat540_2014/examples/photoRec/data/GSE4051_design.rds")
str(prDes)
```

Two sample tests -- one gene
----------------------------

Pick a single gene, somewhat randomly:

```{r}
set.seed(987)
(theGene <- sample(1:nrow(prDat), 1))
```

Make a new data frame, using the field descriptions, plus the gene expression values for the selected gene:

```{r}
pDat <- data.frame(prDes, gExp = unlist(prDat[theGene, ])) 
str(pDat)
```

Calculate the means for each genotype:

```{r}
aggregate(gExp ~ gType, pDat, FUN = mean)
```

A stripplot to sanity-test the t-test result:

```{r}
stripplot(gType ~ gExp, pDat)
```

Next, do a simple t-test:

```{r}
t.test(gExp ~ gType, pDat)
```

Then, save the object to inspect it:

```{r}
ttRes <- t.test(gExp ~ gType, pDat)
str(ttRes)
```

Extract some useful components:

```{r}
ttRes$statistic
ttRes$p.value
```

*Next: You try: draw a different gene at random or pick one for biological interest and look up the Affy probe ID. Use the t test, with and without the common variance assumption, the Wilcoxon, and/or the Kolmogorov-Smirnov test to assess differential expression. Can you pull test statistics and/or p-values from the different approaches into an common object, like a readable table? Are you getting the same message from the various approaches? Hint: wilcox.test(), ks.test().*

