STAT540 Seminar 03 - lattice and ggplot2 graphics
=================================================

Rod Docking - 2014-01-23

This is a workthrough of the material for the [lattice](http://www.ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-lattice.html) and [ggplot2](http://www.ugrad.stat.ubc.ca/~stat540/seminars/seminar03_graphics-ggplot2.html)
tutorials for STAT540.

Load the necessary libraries:

```{r}
library(lattice)
library(ggplot2)
```

Read in the input data set. 
(*Note that this is reading from a symlink to to the original data file from the course repo*)

```{r}
kDat <- readRDS("../../../stat540_2014/examples/photoRec/data/GSE4051_MINI.rds")
str(kDat)
table(kDat$devStage)
table(kDat$gType)
with(kDat, table(devStage, gType))
```

Scatterplots
------------

Make a scatterplot of the expression of two separate genes:

```{r}
xyplot(eggBomb ~ crabHammer, kDat)
```

> You try: request a scatterplot of the variable poisonFang against crabHammer.

OK:

```{r}
xyplot(poisonFang ~ crabHammer, kDat)
```

> Let's imagine that crabHammer is somehow a natural explanatory variable or predictor (weird here, I admit, but go with me) and eggBomb and poisonFang are natural response variables. We might want to see both responses plotted against crabHammer at the same time. Here is a first way to do so, using a bit of a cheat known as the "extended formula interface" in lattice.

```{r}
xyplot(eggBomb + poisonFang ~ crabHammer, kDat,
       auto.key = TRUE)
```

> What if we want each response to have it's own scatterplot, but we want to put them side-by-side for comparison?

```{r}
xyplot(eggBomb + poisonFang ~ crabHammer, kDat,
       outer = TRUE, grid = TRUE)
```

> What if we'd like to know which points are from the wild type mice versus the Nrl knockouts?

```{r}
xyplot(eggBomb + poisonFang ~ crabHammer, kDat,
       outer = TRUE, grid = TRUE,
       groups = gType, auto.key = TRUE)
```

Reshape the data into a better format:

```{r}
nDat <-
    with(kDat,
         data.frame(sidChar, sidNum, devStage, gType, crabHammer,
                    probeset = factor(rep(c("eggBomb", "poisonFang"), each = nrow(kDat))),
                    geneExp = c(eggBomb, poisonFang)))
str(nDat)
```

> Now we can make the previous plot with more canonical lattice syntax, i.e. this workflow and way of thinking will serve you better in the future:

(*Note - above, we've put eggBomb and poisonFang into a new varaible, geneExpe*)

```{r}
xyplot(geneExp ~ crabHammer | probeset, nDat,
       grid = TRUE,
       groups = gType, auto.key = TRUE)
```

> You try: Remake this plot but instead of conveying genotype via color, show developmental stage.

OK:

```{r}
xyplot(geneExp ~ crabHammer | probeset, nDat,
       grid = TRUE,
       groups = devStage, auto.key = TRUE)
```

Stripplot
---------

> The next set of figures we will make requires yet more data reshaping, which is a substantial background task in many analyses. We drop the idea of crabHammer being a predictor and eggBomb and poisonFang being responses and we just treat them all equivalently.

```{r}
oDat <-
    with(kDat,
         data.frame(sidChar, sidNum, devStage, gType,
                    probeset = factor(rep(c("crabHammer", "eggBomb",
                    "poisonFang"), each = nrow(kDat))),
                    geneExp = c(crabHammer, eggBomb, poisonFang)))
str(oDat)
```

> A stripplot is a univariate scatterplot. Let's inspect the gene expression data, plain and simple.

```{r}
stripplot(~ geneExp, oDat)
```

> Pretty boring and slightly nonsensical! We had to start somewhere. Let's split things out for the three different genes.

```{r}
stripplot(probeset ~ geneExp, oDat)
```

> Sometimes, it can help to add jitter, a small bit of meaningless noise, in the horizontal position.

```{r}
stripplot(probeset ~ geneExp, oDat, jitter.data = TRUE)
```

> Notice that all the data is presented in one panel but with the different genes corresponding to different locations in the y direction. What if we want to put the different genes in different panels?

```{r}
stripplot(~ geneExp | probeset, oDat,
          layout = c(nlevels(oDat$probeset), 1))
```

> What if we want to see information about wild type versus Nrl knockout?

```{r}
stripplot(~ geneExp| probeset, oDat, 
          layout = c(nlevels(oDat$probeset), 1),
          groups = gType, auto.key = TRUE)
```

> Let's start exploring gene expression changes over the course of development.

```{r}
stripplot(geneExp ~ devStage, oDat)
```

> Retaining one panel per gene ....

```{r}
stripplot(geneExp ~ devStage | probeset, oDat, 
          layout = c(nlevels(oDat$probeset), 1))
```

> Adding back the genotype information ....

```{r}
stripplot(geneExp ~ devStage | probeset, oDat, 
          layout = c(nlevels(oDat$probeset), 1),
          groups = gType, auto.key = TRUE)
```

> Adding averages

```{r}
stripplot(geneExp ~ devStage | probeset, oDat, 
          layout = c(nlevels(oDat$probeset), 1),
          groups = gType, auto.key = TRUE, grid = TRUE,
          type = c('p', 'a'))
```

> The argument 'type' can be used to add a variety of enhancements. Type is specified as a vector (through the use of 'c'). The option 'p' in the above example specifies the data as points on the plot, 'a' refers to getting the average of each category and joining them by a line (other summaries can be requested too). Some of the other options include 'l' for joining points by lines, 'b' for both points and lines, 'r' for adding the fit from a simple linear regression and 'smooth' for adding a nonparametric "smooth" fitted curve.

Densityplot
-----------

> Here's a nice alternative to histograms!

```{r}
densityplot(~ geneExp, oDat)
```

(*Note - the stripplot of points seems to be part of the default here*)

> The vertical bar works as usual.

```{r}
densityplot(~ geneExp | gType, oDat,
            grid = TRUE)
```

> `groups` works as usual -- a real advantage over histogram.

```{r}
densityplot(~ geneExp, oDat,
            groups = gType, auto.key = TRUE)
```

> The argument 'bw' specifies the bandwidth or the spread of the underlying Gaussian distributions. It controls how smooth this smoothed histogram will be. Though densityplot() has a sensible default, you can always specify directly if you wish. The argument 'n' controls the number of points at which the kernel density estimate is evaluated. It is easy to confuse this with the usual use of 'n' to denote sample size, so beware. If your density looks jaggedy, try increasing 'n'.

```{r}
jBw <- 0.2
jn <- 400
densityplot(~ geneExp, oDat,
            groups = gType, auto.key = TRUE,
            bw = jBw, n = jn,
            main = paste("bw =", jBw, ", n =", jn))
```

> You try: use densityplot() to explore the gene expression distribution by gene and/or developmental stage. Play with 'bw' and 'n' if you like.

Try colouring by gene:

```{r}
densityplot(~ geneExp, oDat,
            groups = probeset, auto.key = TRUE)
```

And then by developmental stage:

```{r}
densityplot(~ geneExp, oDat,
            groups = devStage, auto.key = TRUE)
```

Try splitting into panels (colour is devStage, panels are split by probe):

```{r}
densityplot(~ geneExp | probeset, oDat,
            groups = devStage, auto.key = TRUE)
```

Boxplot
-------

> There is also a time and place for boxplots, obtained with the lattice function bwplot() for "box-and-whiskers plot".
```{r}
bwplot(geneExp ~ devStage, oDat)
```

> The vertical bar | still works ....

```{r}
bwplot(geneExp ~ devStage | gType, oDat)
```

> A violinplot is a hybrid of densityplot and histogram.

```{r}
bwplot(geneExp ~ devStage, oDat,
       panel = panel.violin)
```

Heatmaps
--------

Load in the larger dataset:

```{r}
prDat <- read.table("../../../stat540_2014/examples/photoRec/data/GSE4051_data.tsv")
str(prDat, max.level = 0)
```

Load in the RDS file instead to get the factor levels:

```{r}
prDes <- readRDS("../../../stat540_2014/examples/photoRec/data/GSE4051_design.rds")
str(prDes)
```

> Let's draw 50 probesets at random -- but in a repeatable way!

```{r}
set.seed(1)
(yo <- sample(1:nrow(prDat), size = 50))
hDat <- prDat[yo, ]
str(hDat)
```

> The functions for heatmapping expect a matrix not a data.frame, so we will convert hDat and also transpose for a nicer heatmap orientation below. I also give the samples more informative names that capture genotype and developmental stage.

```{r}
hDat <- as.matrix(t(hDat))
rownames(hDat) <- with(prDes,
                       paste(devStage, gType, sidChar, sep="_"))
str(hDat)
```

Plot a heatmap:

```{r}
heatmap(hDat, Rowv = NA, Colv = NA, scale="none", margins = c(5, 8))
```

> Some of the other built-in color schemes aren't quite as likely to make your eyes bleed ...

```{r}
heatmap(hDat, Rowv = NA, Colv = NA, col = cm.colors(256),
        scale="none", margins = c(5, 8))
```

(_Note: the col argument is the one that changed the colours_)

Load RColorBrewer to get access to more palettes:

```{r}
library(RColorBrewer)
display.brewer.all()
```

Use colorRampPalette to make custom palettes:

```{r}
jGraysFun <- colorRampPalette(brewer.pal(n = 9, "Greys"))
jBuPuFun <- colorRampPalette(brewer.pal(n = 9, "BuPu"))
```

Re-plot the heatmap using the two new palettes:

```{r}
heatmap(hDat, Rowv = NA, Colv = NA, scale="none", margins = c(5, 8),
        col = jGraysFun(256))
heatmap(hDat, Rowv = NA, Colv = NA, scale="none", margins = c(5, 8),
        col = jBuPuFun(256))
```

> By specifying Rowv = NA, Colv = NA, scale = "none", we have been suppressing some rather common heatmap features -- the inclusion of row and column dendrograms and the normalization of the data. Let's look at the heatmap as it would be rendered by default.

```{r}
heatmap(hDat, margins = c(5, 8), col = jBuPuFun(256))
```

> Now we allow scaling within column:

```{r}
heatmap(hDat, col = jBuPuFun(256), margins = c(5, 8), scale=c("column"))
```

> Finally we try out another popular heatmapping function heatmap.2() from the gplots package. This adds an automatic color legend, which helps you determine what each color extreme actually means. If you need to install the package do this:

```{r}
install.packages("gplots")
library(gplots)
heatmap.2(hDat, col = jGraysFun, trace = "none")
heatmap.2(hDat, col = jBuPuFun, trace = "none")
```

Overplotting
------------

> Now that we've loaded the main dataset we can also explore high-volume scatterplotting and the solutions to overplotting. First let's pick two samples at random to plot against each other.

```{r}
set.seed(924)
(yo <- sample(1:ncol(prDat), size = 2))
y <- prDat[[yo[1]]]
z <- prDat[[yo[2]]]
str(y)
str(z)
```

(_Note: subsetting notation to get the two vectors out of the larger object_)

```{r}
xyplot(y ~ z, asp = 1)
```

Overplotting!

```{r}
smoothScatter(y ~ z, asp = 1)
```

> You can see that we were missing some information in the dark cloud above. There is one main clump of data, concentrated around (6, 6) and then petering out diagonally up the x = y line. There's arguably a second, smaller clump of data on a steeper line running through the points ~(10, 8) and ~(14, 14).

> The xyplot() function in lattice can produce a similar plot by specifying a smoothScatter-type of panel function.

```{r}
xyplot(y ~ z, asp = 1, panel = panel.smoothScatter, nbin = 150)
```

> The add-on package hexbin implements hexagonal binning. Basically the plane is divided into hexagons and shaded as described above. Install it if you need to.

```{r}
install.packages("hexbin")
library(hexbin)
hexbinplot(y ~ z)
```

Plot matrix
-----------

Functions for plotting pairs of samples against eachother

```{r}
set.seed(3)
(yo <- sample(1:ncol(prDat), size = 4))
pairDat <- subset(prDat, select = yo)
str(pairDat)
```

> Using the base function pairs() ... You will notice this is a bit slow and we get the usual awful dark point clouds.

```{r}
pairs(pairDat)
```

> However, pairs() can be combined with smoothScatter() for a better result. Somewhat faster and definitely better looking, more informative.

```{r}
pairs(pairDat,
      panel = function(...) smoothScatter(..., add=TRUE))
```

(_Note: I'm not sure what the '...' notation means still_)

> Here's splom() from lattice, first using the default, non-high-volume panel function.

```{r}
splom(pairDat)
```

> Here's splom() from lattice again, but using a smoothScatter-type panel function. Much faster! More informative!

```{r}
splom(pairDat, panel = panel.smoothScatter, raster = TRUE)
```

> Finally, here's hexplom().

```{r}
hexplom(pairDat)
```

Take-home Problem
-----------------

> The full photoRec dataset has 39 samples and 29949 probesets. Choose 2 ... or 20 ... or 200 random probesets/genes and look for gene expression differences between the two genotypes, wild type versus knockout. Make use of the graphing techniques discussed this week such as scatter plots, data heatmaps, correlation heatmaps, etc. Share questions, success, failure on the Google group.



